manta-trusted-setup/src/groth16/ceremony/signature.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Groth16 Trusted Setup Ceremony Signatures
+
+use alloc::vec::Vec;
+use manta_crypto::signature;
+use manta_util::{serde::Serialize, AsBytes};
+
+/// Nonce
+pub trait Nonce: PartialEq {
+    /// Increments the current nonce by one.
```

Comment:
What would be the documentation change? We don't want `Nonce` to be automatically defined for existing types like those that implement `Add` since we need to take care that it satisfies the semantics of a nonce. It's safer to not rely on an existing interface.

Reply:
In this case, we should actually consider instead actually an interface like

```rust
trait Nonce {
    fn increment(&self) -> Self;
    fn are_equal_and_valid(&self, rhs: &Self) -> bool;
}
```

so that there's no doubt. _cc: @BoyuanFeng_

Reply:
1. We may not need `are_equal_and_valid` since we do not always check equality when we check validity.
2. `Add` trait enables addition with any number. We only need an increment of 1 for the nonce.